lib/src/local_working_copy.rs

```diff
@@ -1218,6 +1212,44 @@ impl FileSnapshotter<'_> {
         Ok(())
     }
 
+    /// Emits file paths that don't exist in the `present_entries`.
+    fn emit_deleted_files(
+        &self,
+        dir: &RepoPath,
+        file_states: FileStates<'_>,
+        present_entries: &PresentDirEntries,
+    ) {
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        enum Kind {
+            Dir,
+            File,
+        }
+
+        let file_state_chunks = file_states.iter().chunk_by(|(path, _state)| {
+            // Extract <name> from <dir>, <dir>/<name>, or <dir>/<name>/**.
+            // (file_states may contain <dir> file on file->dir transition.)
+            debug_assert!(path.starts_with(dir));
+            let len = dir.as_internal_file_string().len() + (!dir.is_root() as usize);
+            let tail = path.as_internal_file_string().get(len..).unwrap_or("");
+            match tail.split_once('/') {
+                Some((name, _)) => (Kind::Dir, name),
+                None => (Kind::File, tail),
+            }
```

Comment:
question: Is there a specific invariant here like that the file path has at most one slash in it at this point?

Reply:
The path is normalized `RepoPath`, so it never starts/ends with `/`.